samples/daily-prebuilt/permissions-can-admin/src/components/DailyContainer/DailyContainer.js

```diff
@@ -0,0 +1,284 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { useSearchParams } from 'next/navigation';
+import DailyIframe from '@daily-co/daily-js';
+import JoinForm from '../JoinForm/JoinForm';
+import AdminPanel from '../AdminPanel/AdminPanel';
+import api from '../../daily';
+import './daily-container.css';
+
+export default function DailyContainer() {
+  const searchParams = useSearchParams();
+  const containerRef = useRef(null);
+  const [callFrame, setCallFrame] = useState(null);
+  const [url, setUrl] = useState(null);
+  const [isAdmin, setIsAdmin] = useState(false);
+  const [isOwner, setIsOwner] = useState(false);
+  const [error, setError] = useState(false);
+  const [submitting, setSubmitting] = useState(false);
+  const [participants, setParticipants] = useState({});
+  const prevParticipants = useRef();
+
+  useEffect(() => {
+    prevParticipants.current = participants;
+  }, [participants]);
+
+  useEffect(() => {
+    const urlParam = searchParams.get('url');
+    if (urlParam) {
+      setUrl(urlParam);
+    }
+  }, [searchParams]);
+
+  const handleJoinedMeeting = (e) => {
+    console.log(e.action);
+    setParticipants((p) => ({
+      ...p,
+      [e.participants.local.session_id]: e.participants.local,
+    }));
+  };
+
+  const handleParticipantJoined = (e) => {
+    console.log(e.action);
+    setParticipants((p) => ({
+      ...p,
+      [e.participant.session_id]: e.participant,
+    }));
+  };
+
+  const handleParticipantUpdate = (e) => {
+    console.log(e.action);
+    // // Return early if the participant list isn't set yet.
+    // // This event is sometimes emitted before the joined-meeting event.
+    if (!prevParticipants.current[e.participant.session_id]) return;
```

Comment:
Suggest assigning `e.participant.session_id` as a const up above instead of re-retrieving the nested properties multiple times within these next few lines.